### ISSUE TITLE
Schemagen does not work with indirect inheritance

### ISSUE THREAD

**Issue by MateusMendesSantana** [colyseus/schema#46]
Place.cs will not be generated
```
import { type, MapSchema, ArraySchema } from "@colyseus/schema";
import { Avatar } from "./Avatar";
import { IPlace } from "../models/Place";
import { PlaceMobi } from "./PlaceMobi";
import { VectorTriple } from "./VectorTriple";
import { Quad } from "./Quad";
import { BaseSchema } from "./BaseSchema";

export class Place extends BaseSchema<IPlace> {
  @type("string")
  id: string;
  @type("string")
  name: string;
  @type("string")
  ownerId: string;
  @type("string")
  description: string;
  @type("string")
  rating: number;
  @type("number")
  maxAvatars = 10;
  @type(VectorTriple)
  entrance: VectorTriple;
  @type("number")
  entranceDirection: number;
  @type({ map: Avatar })
  avatars: MapSchema<Avatar>;
  @type([PlaceMobi])
  mobis: ArraySchema<PlaceMobi>;
  @type([Quad])
  quads: ArraySchema<Quad>;

  setData(data: IPlace): void {
    // ..........
  }

  getData() {
    // ...................
  }
}
```
```
import { Schema } from "@colyseus/schema";
import { IParcelable } from "../patterns/IParcelable";

export abstract class BaseSchema<T> extends Schema implements IParcelable<T> {
  constructor(data?: T) {
    super();

    if (data != null) {
      this.setData(data);
    }
  }

  abstract setData(data: T): void;
  abstract getData(): T;
}

```

**Comment by endel** [comment]
Hi @MateusMendesSantana, how're you calling the `schema-codegen` to generate these files? I believe if you provide all schema files in a single `schema-codegen` call, it should generate the `Place.cs` file properly.

Example:

```
schema-codegen ./path/to/structures/*.ts --csharp --output ./cs/directory/
```

Let me know if that works for you, cheers!

**Comment by MateusMendesSantana** [comment]
hi @endel , at first I tried to use * .ts to generate the schema of all models, but it didn't work for me, so believing it was not possible I created my own inscript:
```
const { exec } = require("child_process");
const fs = require("fs");

const source = "./schemas";
const output = "../smadol-client/Assets/Scripts/Schemas";

fs.readdir(source, (error, files) => {
  console.log(files);

  files.forEach(filename => {
    exec(`schema-codegen ${source}/${filename} --csharp --output ${output}`);
  });
});

```
I tried to use the form that guided me, but got the following result (PS: directory "schemas" exists inside "smadol-server"):
```
E:\unity\smadol-server>npx schema-codegen ./schemas/*.ts --csharp --output ../smadol-client/Assets/Scripts/Schemas
internal/fs/utils.js:220
    throw err;
    ^

Error: ENOENT: no such file or directory, open './schemas/*.ts'
?[90m    at Object.openSync (fs.js:440:3)?[39m
?[90m    at Object.readFileSync (fs.js:342:35)?[39m
    at E:\unity\smadol-server\node_modules\?[4m@colyseus?[24m\schema\lib\codegen\parser.js:87:61
    at Array.forEach (<anonymous>)
    at Object.parseFiles (E:\unity\smadol-server\node_modules\?[4m@colyseus?[24m\schema\lib\codegen\parser.js:86:15)
    at Object.<anonymous> (E:\unity\smadol-server\node_modules\?[4m@colyseus?[24m\schema\lib\codegen\cli.js:52:24)
?[90m    at Module._compile (internal/modules/cjs/loader.js:956:30)?[39m
?[90m    at Object.Module._extensions..js (internal/modules/cjs/loader.js:973:10)?[39m
?[90m    at Module.load (internal/modules/cjs/loader.js:812:32)?[39m
?[90m    at Function.Module._load (internal/modules/cjs/loader.js:724:14)?[39m {
  errno: ?[33m-4058?[39m,
  syscall: ?[32m'open'?[39m,
  code: ?[32m'ENOENT'?[39m,
  path: ?[32m'./schemas/*.ts'?[39m
}
```
![image](https://user-images.githubusercontent.com/35671729/67636269-9dae7e80-f8ad-11e9-9724-3b4926fe410c.png)

**Comment by endel** [comment]
Hi @MateusMendesSantana, that's odd, it may be related to the way Windows forward the arguments in the commandline. Can you try providing all the list of files, instead of using `./schemas/*.ts`?

Example:

```
npx schema-codegen schemas/Avatar.ts schemas/BaseSchema.ts schemas/Mobi.ts schemas/MobiCategory.ts schemas/Place.ts  schemas/PlaceMobi.ts schemas/Quad.ts schemas/VectorTriple.ts --csharp --output ../smadol-client/Assets/Scripts/Schemas
```

Let me know if that works for you!

**Comment by MateusMendesSantana** [comment]
@endel, tested here, generated even the second file that directly inherits Schema, in the file Mobi.ts it stopped generating
```
E:\unity\smadol-server>npx schema-codegen schemas/Avatar.ts schemas/BaseSchema.ts schemas/Mobi.ts schemas/MobiCategory.ts schemas/Place.ts  schemas/PlaceMobi.ts schemas/Quad.ts schemas/VectorTriple.ts --csharp --output ../smadol-client/Assets/Scripts/Schemas
[
  'schemas/Avatar.ts',
  'schemas/BaseSchema.ts',
  'schemas/Mobi.ts',
  'schemas/MobiCategory.ts',
  'schemas/Place.ts',
  'schemas/PlaceMobi.ts',
  'schemas/Quad.ts',
  'schemas/VectorTriple.ts'
]
generated: Avatar.cs
generated: BaseSchema.cs
```
**Mobi.ts**
```
import { type, ArraySchema } from "@colyseus/schema";
import { IMobi } from "../models/Mobi";
import { VectorTriple } from "./VectorTriple";
import { MobiCategory } from "./MobiCategory";
import { BaseSchema } from "./BaseSchema";

export class Mobi extends BaseSchema<IMobi> {
  @type("string")
  id: string;
  @type("string")
  name: string;
  @type("boolean")
  walkable: boolean;
  @type("boolean")
  sitable: boolean;
  @type("string")
  description: string;
  @type("int32")
  price: number;
  @type(VectorTriple)
  size: VectorTriple;
  @type(["string"])
  categorieIds: ArraySchema<string>;
  @type([MobiCategory])
  categories: ArraySchema<MobiCategory>;

  setData(data: IMobi): void {
    this.id = data.id;
    this.name = data.name;
    this.description = data.description;
    this.walkable = data.walkable;
    this.sitable = data.sitable;
    this.price = data.price;
    this.size = new VectorTriple(data.size);
    this.categorieIds = new ArraySchema(...data.categorieIds.map(it => it.toString()));
    this.categories = new ArraySchema(...data.categories.map(it => new MobiCategory(it)));
  }

  getData() {
    return {
      id: this.id,
      name: this.name,
      description: this.description,
      walkable: this.walkable,
      sitable: this.sitable,
      price: this.price,
      size: this.size,
      categorieIds: this.categorieIds,
      categories: this.categories.map(it => it.getData())
    } as unknown as IMobi;
  }
}
```
**BaseSchema.ts**
```

import { Schema } from "@colyseus/schema";
import { IParcelable } from "../patterns/IParcelable";

export abstract class BaseSchema<T> extends Schema implements IParcelable<T> {
  constructor(data?: T) {
    super();

    if (data != null) {
      this.setData(data);
    }
  }

  abstract setData(data: T): void;
  abstract getData(): T;
}

```

**Comment by endel** [comment]
That's odd @MateusMendesSantana, do you mind sharing a reproductible scenario where I can see this issue happening? If you can create a public repo on GitHub with your schemas it would be great.

I've added a new test case for testing codegen here: https://github.com/colyseus/schema/blob/master/test/codegen/CodegenTest.ts#L19-L29

The test is getting all the files from `test/codegen/sources/*.ts` files and generating their `.cs` version under `test/codegen/sources/tmp-codegen-output` (which is later removed after tests run)

If you can reproduce the issue by running the test would be perfect. Here's the command to run only the codegen tests:

```
npm test -- --grep 'schema-codegen'
```

Cheers, have a nice week!

**Comment by MateusMendesSantana** [comment]
hi @endel. I run the test script, so far so good:
![image](https://user-images.githubusercontent.com/35671729/67700406-aa55d400-f98c-11e9-8d60-7cfa76b655df.png)
So I came up with the idea of turning BaseSchema into a generic class:
```
import { Schema, type } from "../../../src";

export class BaseSchema<T = any> extends Schema {
    @type("number") id: number;
}
```
And add **< any >** in the **Inheritance2** class:
```
import { type } from "../../../src";
import { BaseSchema } from "./BaseSchema";

class Inheritance2 extends BaseSchema<any> {
  @type("number") x: number;
  @type("number") y: number;
}
```
![image](https://user-images.githubusercontent.com/35671729/67705584-053ff900-f996-11e9-9e28-bfc82d01f2e3.png)

**Comment by endel** [comment]
Awesome @MateusMendesSantana, thanks a lot for your insights! This has been fixed now on `@colyseus/schema@0.5.2` 🎉 

I've managed to reproduce using the generics you've mentioned, and pushed a fix for it. If you remove your `node_modules` + `package-lock.json` and install them again, it should be fine now! 

Let me know if you find any other issues! Cheers!
